slither/utils/upgradeability.py

```diff
@@ -0,0 +1,463 @@
+from slither.core.cfg.node import Node, NodeType
+from slither.slithir.operations import (
+    Assignment,
+    Index,
+    Member,
+    Length,
+    Binary,
+    Unary,
+    Condition,
+    NewArray,
+    NewStructure,
+    NewContract,
+    NewElementaryType,
+    SolidityCall,
+    Delete,
+    EventCall,
+    LibraryCall,
+    InternalDynamicCall,
+    HighLevelCall,
+    LowLevelCall,
+    TypeConversion,
+    Return,
+    Transfer,
+    Send,
+    Unpack,
+    InitArray,
+    InternalCall,
+)
+from slither.slithir.variables import (
+    TemporaryVariable,
+    TupleVariable,
+    Constant,
+    ReferenceVariable,
+)
+from slither.tools.read_storage.read_storage import SlotInfo, SlitherReadStorage
+
+
+# pylint: disable=too-many-locals
+def compare(v1: Contract, v2: Contract) -> dict:
+    """
+    Compares two versions of a contract. Most useful for upgradeable (logic) contracts,
+    but does not require that Contract.is_upgradeable returns true for either contract.
+
+    Args:
+        v1: Original version of (upgradeable) contract
+        v2: Updated version of (upgradeable) contract
+
+    Returns: dict {
+        "missing-vars-in-v2": list[Variable],
+        "new-variables": list[Variable],
+        "tainted-variables": list[Variable],
+        "new-functions": list[Function],
+        "modified-functions": list[Function],
+        "tainted-functions": list[Function]
+    }
+    """
+
+    order_vars1 = [v for v in v1.state_variables if not v.is_constant and not v.is_immutable]
+    order_vars2 = [v for v in v2.state_variables if not v.is_constant and not v.is_immutable]
+    func_sigs1 = [function.solidity_signature for function in v1.functions]
+    func_sigs2 = [function.solidity_signature for function in v2.functions]
+
+    results = {
+        "missing-vars-in-v2": [],
+        "new-variables": [],
+        "tainted-variables": [],
+        "new-functions": [],
+        "modified-functions": [],
+        "tainted-functions": [],
+    }
+
+    # Since this is not a detector, include any missing variables in the v2 contract
+    if len(order_vars2) < len(order_vars1):
+        for variable in order_vars1:
+            if variable.name not in [v.name for v in order_vars2]:
+                results["missing-vars-in-v2"].append(variable)
+
+    # Find all new and modified functions in the v2 contract
+    new_modified_functions = []
+    new_modified_function_vars = []
+    for sig in func_sigs2:
+        function = v2.get_function_from_signature(sig)
+        orig_function = v1.get_function_from_signature(sig)
+        if sig not in func_sigs1:
+            new_modified_functions.append(function)
+            results["new-functions"].append(function)
+            new_modified_function_vars += (
+                function.state_variables_read + function.state_variables_written
+            )
+        elif not function.name.startswith("slither") and is_function_modified(
```

Comment:
I think here we should use `not function. is_constructor_variables` instead of comparing the name

https://github.com/crytic/slither/blob/master/slither/core/declarations/function.py#L396-L404